File: java/fesql-spark/src/main/scala/com/_4paradigm/hybridse/spark/utils/FesqlUtil.scala
/*
 * Copyright 2021 4Paradigm
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com._4paradigm.hybridse.spark.utils

import java.util

import com._4paradigm.hybridse.`type`.TypeOuterClass.{ColumnDef, Database, TableDef, Type}
import com._4paradigm.hybridse.vm.PhysicalOpNode
import com._4paradigm.hybridse.node.{DataType => InnerDataType}
import org.apache.spark.sql.{DataFrame, Row}
import org.apache.spark.sql.types._

import scala.collection.JavaConverters._


object FesqlUtil {

  def getOutputSchemaSlices(node: PhysicalOpNode): Array[StructType] = {
    (0 until node.GetOutputSchemaSourceSize().toInt).map(i => {
      val columnDefs = node.GetOutputSchemaSource(i).GetSchema()
      getSparkSchema(columnDefs)
    }).toArray
  }

  def getDatabase(databaseName: String, dict: Map[String, DataFrame]): Database = {
    val databaseBuilder = Database.newBuilder()
    databaseBuilder.setName(databaseName)
    dict.foreach { case (name, df) =>
      databaseBuilder.addTables(getTableDef(name, df))
    }
    databaseBuilder.build()
  }

  def getTableDef(tableName: String, dataFrame: DataFrame): TableDef = {
    val tblBulder = TableDef.newBuilder()
    dataFrame.schema.foreach(field => {
      tblBulder.addColumns(ColumnDef.newBuilder()
        .setName(field.name)
        .setIsNotNull(! field.nullable)
        .setType(getFeSQLType(field.dataType))
        .build()
      )
    })
    tblBulder.setName(tableName)
    tblBulder.build()
  }

  def getFeSQLSchema(structType: StructType): java.util.List[ColumnDef] = {
    val list = new util.ArrayList[ColumnDef]()
    structType.foreach(field => {
      list.add(ColumnDef.newBuilder()
        .setName(field.name)
        .setIsNotNull(!field.nullable)
        .setType(getFeSQLType(field.dataType)).build())
    })
    list
  }

  def getSparkSchema(columns: java.util.List[ColumnDef]): StructType = {
    StructType(columns.asScala.map(col => {
      StructField(col.getName, getSparkType(col.getType), !col.getIsNotNull)
    }))
  }

  def getSparkType(dtype: Type): DataType = {
    dtype match {
      case Type.kInt16 => ShortType
      case Type.kInt32 => IntegerType
      case Type.kInt64 => LongType
      case Type.kFloat => FloatType
      case Type.kDouble => DoubleType
      case Type.kBool => BooleanType
      case Type.kVarchar => StringType
      case Type.kDate => DateType
      case Type.kTimestamp => TimestampType
      case _ => throw new IllegalArgumentException(
        s"FeSQL type $dtype not supported")
    }
  }

  def getFeSQLType(dtype: DataType): Type = {
    dtype match {
      case ShortType => Type.kInt16
      case IntegerType => Type.kInt32
      case LongType => Type.kInt64
      case FloatType => Type.kFloat
      case DoubleType => Type.kDouble
      case BooleanType => Type.kBool
      case StringType => Type.kVarchar
      case DateType => Type.kDate
      case TimestampType => Type.kTimestamp
      case _ => throw new IllegalArgumentException(
        s"Spark type $dtype not supported")
    }
  }

  def getSchemaTypeFromInnerType(dtype: InnerDataType): Type = {
    dtype match {
      case InnerDataType.kInt16 => Type.kInt16
      case InnerDataType.kInt32 => Type.kInt32
      case InnerDataType.kInt64 => Type.kInt64
      case InnerDataType.kFloat => Type.kFloat
      case InnerDataType.kDouble => Type.kDouble
      case InnerDataType.kBool => Type.kBool
      case InnerDataType.kVarchar => Type.kVarchar
      case InnerDataType.kDate => Type.kDate
      case InnerDataType.kTimestamp => Type.kTimestamp
      case _ => throw new IllegalArgumentException(
        s"Inner type $dtype not supported")
    }
  }

  def getInnerTypeFromSchemaType(dtype: Type): InnerDataType = {
    dtype match {
      case Type.kInt16 => InnerDataType.kInt16
      case Type.kInt32 => InnerDataType.kInt32
      case Type.kInt64 => InnerDataType.kInt64
      case Type.kFloat => InnerDataType.kFloat
      case Type.kDouble => InnerDataType.kDouble
      case Type.kBool => InnerDataType.kBool
      case Type.kVarchar => InnerDataType.kVarchar
      case Type.kDate => InnerDataType.kDate
      case Type.kTimestamp => InnerDataType.kTimestamp
      case _ => throw new IllegalArgumentException(
        s"Schema type $dtype not supported")
    }
  }

  def createGroupKeyComparator(keyIdxs: Array[Int], schema: StructType): (Row, Row) => Boolean = {
    if (keyIdxs.length == 1) {
      val idx = keyIdxs(0)
      (row1, row2) => {
        row1.get(idx) != row2.get(idx)
      }
    } else {
      (row1, row2) => {
        keyIdxs.exists(i => row1.get(i) != row2.get(i))
      }
    }
  }

}

File: java/fesql-flink/src/main/java/com/_4paradigm/hybridse/flink/common/FesqlUtil.java
/*
 * Copyright 2021 4Paradigm
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com._4paradigm.hybridse.flink.common;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com._4paradigm.hybridse.common.FesqlException;
import com._4paradigm.hybridse.node.ColumnRefNode;
import com._4paradigm.hybridse.node.ExprNode;
import com._4paradigm.hybridse.node.ExprType;
import com._4paradigm.hybridse.vm.CoreAPI;
import com._4paradigm.hybridse.vm.PhysicalOpNode;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.typeutils.RowTypeInfo;
import org.apache.flink.table.types.DataType;
import org.apache.flink.table.types.logical.*;
import org.apache.flink.table.types.logical.LogicalType;
import org.apache.flink.table.api.TableColumn;
import org.apache.flink.table.api.TableSchema;
import com._4paradigm.hybridse.type.TypeOuterClass;
import com._4paradigm.hybridse.type.TypeOuterClass.Type;
import static com._4paradigm.hybridse.type.TypeOuterClass.Type.*;


public class FesqlUtil {

    /**
     * Build FESQL datatype with Flink datatype.
     */
    public static Type getFesqlType(DataType flinkDataType) throws FesqlException {
        LogicalType logicalType = flinkDataType.getLogicalType();
        if (logicalType instanceof IntType) {
            // Notice that no short or long in flink logical type
            return Type.kInt32;
        } else if (logicalType instanceof BigIntType) {
            return Type.kInt64;
        } else if (logicalType instanceof FloatType) {
            return Type.kFloat;
        } else if (logicalType instanceof DoubleType) {
            return Type.kDouble;
        } else if (logicalType instanceof VarCharType) {
            return Type.kVarchar;
        } else if (logicalType instanceof DateType) {
            return Type.kDate;
        } else if (logicalType instanceof TimestampType) {
            return Type.kTimestamp;
        } else {
            throw new FesqlException(String.format("Do not support Flink datatype %s", flinkDataType));
        }
    }

    /**
     * Build FESQL table def with Flink table schema.
     */
    public static TypeOuterClass.TableDef buildTableDef(String tableName, TableSchema tableSchema) throws FesqlException {

        TypeOuterClass.TableDef.Builder tableBuilder = TypeOuterClass.TableDef.newBuilder();
        tableBuilder.setName(tableName);

        for (TableColumn tableColumn: tableSchema.getTableColumns()) {
            TypeOuterClass.ColumnDef columnDef = TypeOuterClass.ColumnDef.newBuilder()
                    .setName(tableColumn.getName())
                    .setIsNotNull(!tableColumn.getType().getLogicalType().isNullable())
                    .setType(getFesqlType(tableColumn.getType()))
                    .build();
            tableBuilder.addColumns(columnDef);
        }

        return tableBuilder.build();
    }

    /**
     * Build FESQL database with map of table name and schema.
     */
    public static TypeOuterClass.Database buildDatabase(String dbName, Map<String, TableSchema> tableSchemaMap) throws FesqlException {

        TypeOuterClass.Database.Builder builder = TypeOuterClass.Database.newBuilder();
        builder.setName(dbName);

        for (Map.Entry<String, TableSchema> entry : tableSchemaMap.entrySet()) {
            builder.addTables(buildTableDef(entry.getKey(), entry.getValue()));
        }

        return builder.build();
    }

    /**
     * Get the node output schema as list of slice and the slice is list of column def.
     */
    public static List<List<TypeOuterClass.ColumnDef>> getNodeOutputColumnLists(PhysicalOpNode node) {
        List<List<TypeOuterClass.ColumnDef>> outputLists = new ArrayList<List<TypeOuterClass.ColumnDef>>();

        for (int i=0; i < node.GetOutputSchemaSourceSize(); ++i) {
            List<TypeOuterClass.ColumnDef> columnDefs = node.GetOutputSchemaSource(i).GetSchema();
            outputLists.add(columnDefs);
        }

        return outputLists;
    }

    public static RowTypeInfo generateRowTypeInfo(List<TypeOuterClass.ColumnDef> columnDefs) throws FesqlException {
        int fieldNum = columnDefs.size();
        TypeInformation<?>[] fieldTypes = new TypeInformation<?>[fieldNum];

        for (int i=0; i < fieldNum; ++i) {
            TypeOuterClass.Type columnType = columnDefs.get(i).getType();

            if (columnType == kInt16) {
                fieldTypes[i] = Types.SHORT;
            } else if (columnType == kInt32) {
                fieldTypes[i] = Types.INT;
            } else if (columnType == kInt64) {
                // TODO: Catch TableException: Type is not supported: BigInteger
                fieldTypes[i] = Types.BIG_INT;
                //fieldTypes[i] = Types.INT;
            } else if (columnType== kFloat) {
                fieldTypes[i] = Types.FLOAT;
            } else if (columnType == kDouble) {
                fieldTypes[i] = Types.DOUBLE;
            } else if (columnType == kBool) {
                fieldTypes[i] = Types.BOOLEAN;
            } else if (columnType == kVarchar) {
                fieldTypes[i] = Types.STRING;
            } else if (columnType == kTimestamp) {
                // TODO: Make sure it is compatible with sql timestamp type
                fieldTypes[i] = Types.SQL_TIMESTAMP;
            } else if (columnType == kDate) {
                fieldTypes[i] = Types.SQL_DATE;
            } else {
                throw new FesqlException(String.format("Fail to convert row type info with %s", columnType));
            }

        }

        return new RowTypeInfo(fieldTypes);
    }

    public static int resolveColumnIndex(ExprNode exprNode, PhysicalOpNode physicalOpNode) throws FesqlException {
        if (exprNode.getExpr_type_().swigValue() == ExprType.kExprColumnRef.swigValue()) {
            int index = CoreAPI.ResolveColumnIndex(physicalOpNode, ColumnRefNode.CastFrom(exprNode));
            if (index < 0) {
                throw new FesqlException("Fail to resolve column ref expression node and get index " + index);
            } else if (index >= physicalOpNode.GetOutputSchema().size()) {
                throw new FesqlException("Fail to resolve column ref  expression node and get index " + index);
            } else {
                return index;
            }
        } else {
            throw new FesqlException("Do not support nono-columnref expression");
        }
    }
}
